Add RFC about WASM cache

flavio opened the pull request:
Initial version of the RFC.

flavio committed:
Add RFC about WASM cache

Initial version of the RFC

Signed-off-by: Flavio Castelli <fcastelli@suse.com>

flavio commented:
This is a draft, I didn't have time to proofread it since I have to leave for holidays.

@jvanz / @viccuad: feel free to take a look at that if you have time. As you will see I think there are other approaches that might be easier to accomplish. We can discuss more once I'm back

flavio committed:
WASM cache RFC: add another possible option

Document the new `ReadWriteOncePod` access mode

Signed-off-by: Flavio Castelli <fcastelli@suse.com>

viccuad commented:
LGTM here, the listed alternatives are well documented. From the listed ones, I favour an optional PV and no StatefulSets. I look forward to follow-up discussions.

viccuad commented on rfc/0016-wasm-cache.md:
```diff
@@ -0,0 +1,226 @@
+
+## The caching OCI registry
+
+The OCI registry used for caching would not be exposed outside of the cluster.
```
in case the cluster is airgapped, we could reuse the OCI registry inside of the airgap.

jvanz reviewed (commented):
I'm okay with the proposed approach. But I just had an idea...

What if do we remove the load of pre compiling the policies from the policy server? We could use a job which will download the policy pre-compile it and store in a PV. Therefore, policy servers does not write to it just read it. No write concurrent access. The controller can watch for new policies, triggers this job, and only after the successful job execution update/restart the policy server. This job could use a new `kwctl` command to optimize the policies and write the result into the PV read after by the policy server.

flavio commented:
> What if do we remove the load of pre compiling the policies from the policy server? We could use a job which will download the policy pre-compile it and store in a PV. Therefore, policy servers does not write to it just read it. No write concurrent access. The controller can watch for new policies, triggers this job, and only after the successful job execution update/restart the policy server. This job could use a new `kwctl` command to optimize the policies and write the result into the PV read after by the policy server.

Thanks for the suggestion. I still prefer one of the alternative approaches because this would not complicate the rollout strategy used by the controller to start a new instance of the policy server.

Let's experiment a bit with the kubernetes leases and a shared PV. I'll keep you posted.

flavio committed:
Reject the wasm-cache RFC

The alternative solutions illustrated inside of the RFC provide more
appealing solutions.

Signed-off-by: Flavio Castelli <fcastelli@suse.com>

flavio commented:
Closing the RFC as rejected. The alternative solutions illustrated inside of the RFC are more appealing.

I'm going to open a new RFC to explain how one of the alternative solutions can be adopted to solve this problem

flavio commented:
Sorry, I accidentally closed the PR. I'll instead merge it, as it's supposed to be. The RFC has status `REJECTED`
